### ISSUE TITLE
ALE Fixers - (Javascript)? prettier - how to install and use?

### ISSUE THREAD

**Issue by dfishburn** [w0rp/ale#887]
I wanted to use one of the features of ALE, but "just can't get there from here".
Thought I would take you through my flow to figure out how to use it, so that if documentation updates are required (I think they are), they will help someone else enable these features.

First the usual:
```
:ver
VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Aug  3 2017 08:16:42)
MS-Windows 32-bit GUI version with OLE support
Included patches: 1-846
```
ALEInfoToClipboard
```
 Current Filetype: javascript
Available Linters: ['eslint', 'flow', 'jscs', 'jshint', 'standard', 'xo']
  Enabled Linters: ['eslint']
 Linter Variables:

let g:ale_javascript_eslint_executable = 'eslint'
let g:ale_javascript_eslint_options = ''
let g:ale_javascript_eslint_suppress_eslintignore = 0
let g:ale_javascript_eslint_use_global = 0
 Global Variables:

let g:ale_echo_cursor = 1
let g:ale_echo_msg_error_str = 'Error'
let g:ale_echo_msg_format = '%s'
let g:ale_echo_msg_warning_str = 'Warning'
let g:ale_enabled = 1
let g:ale_fix_on_save = 0
let g:ale_fixers = {'javascript': ['prettier', 'eslint']}
let g:ale_keep_list_window_open = 0
let g:ale_lint_delay = 200
let g:ale_lint_on_enter = 1
let g:ale_lint_on_save = 1
let g:ale_lint_on_text_changed = 'always'
let g:ale_linter_aliases = {}
let g:ale_linters = {'javascript': ['eslint']}
let g:ale_open_list = 0
let g:ale_set_highlights = 1
let g:ale_set_loclist = 1
let g:ale_set_quickfix = 0
let g:ale_set_signs = 1
let g:ale_sign_column_always = 0
let g:ale_sign_error = '>>'
let g:ale_sign_offset = 1000000
let g:ale_sign_warning = '--'
let g:ale_statusline_format = ['%d error(s)', '%d warning(s)', 'OK']
let g:ale_warn_about_trailing_whitespace = 1
  Command History:

(executable check - success) eslint
(finished - exit code 1) 'cmd /c eslint -f unix --stdin --stdin-filename N:\dave.js < C:\Users\i826639\AppData\Local\Temp\VIDA1EC.tmp\dave.js'

<<<OUTPUT STARTS>>>

Oops! Something went wrong! :(

ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    eslint --init

ESLint looked for configuration files in N:\ and its ancestors.

If you think you already have a configuration file or if you need more help, please stop by the ESLint chat room: https://gitter.im/eslint/eslint

<<<OUTPUT ENDS>>>

```

I just installed ALE 1.5.0.
As I was reading through the update, I noticed the section on "Fixers".
Didn't know what those were.

ale.txt

> 4. Fixing Problems                                                    *ale-fix*
> 
> ALE can fix problems with files with the |ALEFix| command. When |ALEFix| is
> run, the variable |g:ale_fixers| will be read for getting a |List| of commands
> for filetypes, split on `.`, and the functions named in |g:ale_fixers| will be
> executed for fixing the errors.

Nothing here really describes what a "fixer" is though.  Do all the fixers do the same thing?  They reformat code?

Googling around I found this link:
[A Prettier JavaScript Formatter](http://jlongster.com/A-Prettier-Formatter)
From January 2017, he gives a pretty good description of what one of the few I would be interested in does.  Specifically, it will intelligently reformat my code **as I write it**.

Looking through ale-javascript.txt I see this section:

> prettier-eslint                                *ale-javascript-prettier-eslint*
> 
> ALE supports `prettier-eslint` for easy integration with projects, but it is
> not recommended for new projects. ALE instead recommends configuring
> |g:ale_fixers| to run `'prettier'` and `'eslint'` in a sequence like so: >
> 
>   let g:ale_fixers = {'javascript': ['prettier', 'eslint']}

So I added the g:ale_fixers to my vimrc and restarted and tried some stuff and nothing happened.

No error messages, no real information that anything was even attempted (and possibly failed).

So I am assuming I don't have the "executable" available.

Googled around and found this nodejs module:
[standard-prettier](https://www.npmjs.com/package/standard-prettier)

Installed it (on Windows 10):
```
$ npm install -g prettier-standard-formatter
C:\Users\i826639\AppData\Roaming\npm\prettier-standard-formatter -> C:\Users\i826639\AppData\Roaming\npm\node_modules\prettier-standard-formatter\cli.js
prettier-standard-formatter@0.222222222222222.333333333333333 C:\Users\i826639\AppData\Roaming\npm\node_modules\prettier-standard-formatter
├── resolve-cwd@1.0.0 (resolve-from@2.0.0)
├── recursive-readdir@2.2.1 (minimatch@3.0.3)
├── globby@6.1.0 (pify@2.3.0, object-assign@4.1.1, pinkie-promise@2.0.1, array-union@1.0.2, glob@7.1.2)
├── meow@3.7.0 (map-obj@1.0.1, decamelize@1.2.0, trim-newlines@1.0.0, object-assign@4.1.1, minimist@1.2.0, camelcase-keys@2.1.0, loud-rejection@1.6.0, normalize-package-data@2.4.0, redent@1.0.0, read-pkg-up@1.0.1)
├── standard@8.6.0 (eslint-config-standard-jsx@3.2.0, eslint-config-standard@6.2.1, eslint-plugin-standard@2.0.1, eslint-plugin-promise@3.4.2, standard-engine@5.2.0, eslint-plugin-react@6.7.1, eslint@3.10.2)
└── prettier@0.19.0 (get-stdin@5.0.1, esutils@2.0.2, babylon@6.15.0, minimist@1.2.0, ast-types@0.9.4, babel-code-frame@6.22.0, chalk@1.1.3, glob@7.1.1, jest-validate@19.0.0, flow-parser@0.38.0)
```

Updated my vimrc to use:
```
if has('channel')
    " Restrict Javascript to only eslint
    let g:ale_linters = {'javascript': ['eslint']}
    " Fixers are automatic code formatters
    " To install:
    "     npm install -g prettier-standard-formatter
    let g:ale_fixers = {'javascript': ['prettier-standard', 'eslint']}

    set runtimepath^=$VIM/vimfiles/vim-ale/vim-ale-1.5.0
endif
```

Restarted Vim.
:e dave.js
Typed in the following code and hit enter at the end:
`foo(reallyLongArg(), omgSoManyParameters(), IShouldRefactorThis(), akjsdfklajsdklfjsklfjklsdafjsdjflajsfkjaslfjkasjfjdsfkljaskfjafjklsd());`

Nothing happened to the code, which I was expecting this to happen:
```
foo(
    reallyLongArg(), 
    omgSoManyParameters(), 
    IShouldRefactorThis(), 
    akjsdfklajsdklfjsklfjklsdafjsdjflajsfkjaslfjkasjfjdsfkljaskfjafjklsd()
);
```

Thought I would need to debug some, and thought I remembered ALE having some gear in place to help with that.
```
:h ale.txt
/\cdebug
```

Didn't find anything.

So, remembered reading about specifying the location of the binary:
$ C:\Users\i826639\AppData\Roaming\npm\node_modules\prettier-standard-formatter\node_modules\.bin\prettier.cmd
Usage: prettier [opts] [filename ...]

Available options:
  --write                  Edit the file in-place. (Beware!)
  --stdin                  Read input from stdin.
  --print-width <int>      Specify the length of line that the printer will wrap on. Defaults to 80.
  --tab-width <int>        Specify the number of spaces per indentation-level. Defaults to 2.
  --single-quote           Use single quotes instead of double.
  --bracket-spacing        Put spaces between brackets. Defaults to true.
  --jsx-bracket-same-line  Put > on the last line. Defaults to false.
  --trailing-comma <none|es5|all>
                           Print trailing commas wherever possible. Defaults to none.
  --parser <flow|babylon>  Specify which parse to use. Defaults to babylon.
  --color                  Colorize error messages. Defaults to true.
  --version                Print prettier version.

Boolean options can be turned off like this:
  --no-bracket-spacing
  --bracket-spacing=false

Set this in my buffer:
`    let g:ale_javascript_prettier_standard_executable = 'C:\Users\i826639\AppData\Roaming\npm\node_modules\prettier-standard-formatter\node_modules\.bin\prettier.cmd'`

Then in my code, retyped that really long test line again, and nothing happened still.

In my vimrc I now have:
```
if has('channel')
    " Restrict Javascript to only eslint
    let g:ale_linters = {'javascript': ['eslint']}
    " Fixers are automatic code formatters
    " To install:
    "     npm install -g prettier-standard-formatter
    let g:ale_fixers = {'javascript': ['prettier-standard', 'eslint']}
    let g:ale_javascript_prettier_standard_executable = 'C:\Users\i826639\AppData\Roaming\npm\node_modules\prettier-standard-formatter\node_modules\.bin\prettier.cmd'

    set runtimepath^=$VIM/vimfiles/vim-ale/vim-ale-1.5.0
endif
```

I would like to get this prettier working with Javascript **and Java (if available)**.

Can anyone help me along.

**Comment by w0rp** [comment]
What?

Just use `npm install prettier` in your project directory to install prettier to `node_modules` and use something like `let g:ale_fixers = {'javascript': ['prettier']}`. It's not rocket science.

**Comment by w0rp** [comment]
If you want to use `eslint`, then read the documentation for `eslint`, which explains how to install ESLint and configure it.

**Comment by dfishburn** [comment]
> It's not rocket science.

Agreed, it shouldn't be.
That is why I showed what I have tried and why and where it got me.
And where the existing documentation got me, what I looked for and where I ended up.

> Just use npm install prettier in your project directory

So that is an interesting comment on it's own.
You mention a "project directory".
I don't use a specific project, I just want this to work with any Javascript file I open regardless of "project"
This is why I used the -g flag:
`npm install -g prettier`

Should this work just fine this way too?

With regard to trying to figure out why it was not working, what can you suggest?

**Comment by w0rp** [comment]
Yeah, installing the packages globally should work too. You can install Node programs either globally or locally.

If you want to ask for more documentation or suggest improvements, you need to collect your thoughts first. I can't really do anything based on your comments above. If you have a specific request, then I can look at that.

**Comment by w0rp** [comment]
ALE isn't responsible for describing how the NPM and JavaScript ecosystem works, but it might be possible to add a sentence or two somewhere in the documentation to eliminate confusion.

**Comment by dfishburn** [comment]
With regard to trying to figure out why it was not working, what can you suggest?

**Comment by habamax** [comment]
@dsifford you have to run fixer.

Prettier to format HTML:

```
let g:ale_fixers = {}
let g:ale_fixers.html = ['prettier']
let g:ale_fixers['*'] = ['remove_trailing_lines', 'trim_whitespace']
let g:ale_javascript_prettier_options = '--use-tabs'
```

Then in vim run `:ALEFix` command for HTML file.

**Comment by gerrard00** [comment]
@dfishburn if you look at the end of the data shown by `ALEInfo` (as you posted) you see the command history as well as the output from those commands. If the output doesn't make the issue immediately clear you can copy paste the command and try running it in the shell directly. 

For example in your original post you see "exit code 1" and "ESLint couldn't find a configuration file" which would tell you how to correct your issues wih eslint. In your case you need to create an eslint configuration file in order to tell eslint what linting rules it should apply. I would recommend getting eslint working by itself at a command line before throwing vim and ALE into the mix.

I was having a similar issue today where prettier was not running. Checking my command output with ALEInfo I could see that the call to prettier was failing. I ran the same command directly in the terminal and prettier showed a very clear error message:

```
[error] Invalid --trailing-comma value. Expected "all", "es5" or "none", but received "always".
```
I my case I had an invalid option set for prettier. I fixed that in my vimrc file and now prettier works as expected.

HTH

TL;DR check the end of the `ALEInfo` output for the command that ALE is trying to execute for you.
